src/gui/app.cpp

```diff
@@ -1,18 +1,150 @@
+		: type{ getConnectorType(parser) }
+		, timeout{ parser.value("timeout").toInt() }
+		, server{ parser.value("server") }
+		, nvim{ parser.value("nvim") }
+		, positionalArgs{ parser.positionalArguments() }
+		, neovimArgs{ std::move(nvimArgs) }
+	{
+	}
+
+	ConnectorInitArgs(
+		Type _type, int _timeout, QString _server, QString _nvim, QStringList nvimArgs)
+		: type{ _type }
+		, timeout{ _timeout }
+		, server{ std::move(_server) }
+		, nvim{ std::move(_nvim) }
+		, positionalArgs{}
+		, neovimArgs{ std::move(nvimArgs) }
+	{
+	}
+
+	const Type type;
+	const int timeout;
+	const QString server;
+	const QString nvim;
+	const QStringList positionalArgs;
+	const QStringList neovimArgs;
+
+	Type getConnectorType(const QCommandLineParser& parser)
+	{
+		if (parser.isSet("server")) {
+			return ConnectorInitArgs::Type::Server;
+		}
+
+		if (parser.isSet("embed")) {
+			return ConnectorInitArgs::Type::Embed;
+		}
+
+		if (parser.isSet("spawn") && !parser.positionalArguments().isEmpty()) {
+			return ConnectorInitArgs::Type::Spawn;
+		}
+
+		return ConnectorInitArgs::Type::Default;
+	}
+};
+
+NeovimConnector* connectToRemoteNeovim(const ConnectorInitArgs& args)
+{
+	NeovimConnector* connector{ nullptr };
+	if (args.type == ConnectorInitArgs::Type::Embed) {
+		connector = NeovimQt::NeovimConnector::fromStdinOut();
+	}
+
+	if (args.type == ConnectorInitArgs::Type::Server) {
+		connector = NeovimQt::NeovimConnector::connectToNeovim(args.server);
+	}
+
+	if (args.type == ConnectorInitArgs::Type::Spawn) {
+		Q_ASSERT(!args.positionalArgs.isEmpty());
+		connector =
+			NeovimQt::NeovimConnector::spawn(args.positionalArgs.mid(1), args.positionalArgs.at(0));
+	}
+
+	if (!connector) {
+		connector =
+			NeovimQt::NeovimConnector::spawn(args.neovimArgs + args.positionalArgs, args.nvim);
+	}
+
+	connector->setRequestTimeout(args.timeout);
+	return connector;
+}
+
+void onWindowActiveChanged(MainWindow& window)
+{
+	s_lastActiveWindow = &window;
+}
+
+void onFileOpenEvent(const QList<QUrl>& files)
+{
+	Q_ASSERT(s_lastActiveWindow);
+	s_lastActiveWindow->shell()->openFiles(files);
+}
+
+void onWindowClosing(int status)
+{
+	s_exitStatus = status;
+}
+
+void onWindowDestroyed()
+{
+	if (s_windows.empty()) {
+		qApp->exit(s_exitStatus);
+	}
+}
+
+MainWindow* createWindow(NeovimConnector* connector)
+{
+	Q_ASSERT(connector);
+	NeovimQt::MainWindow* win = new NeovimQt::MainWindow(connector);
+	win->setAttribute(Qt::WA_DeleteOnClose);
+
+	App* app = qobject_cast<App*>(App::instance());
+	Q_ASSERT(app);
+
+	QObject::connect(win, &MainWindow::closing, app, onWindowClosing);
+	QObject::connect(win, &MainWindow::destroyed, app, [win]() {
+		Q_ASSERT(std::find(s_windows.cbegin(), s_windows.cend(), win) != s_windows.cend());
+		s_windows.erase(std::find(s_windows.cbegin(), s_windows.cend(), win));
+		onWindowDestroyed();
+	});
+	QObject::connect(win, &MainWindow::activeChanged, app, onWindowActiveChanged);
+
+	s_lastActiveWindow = win;
```

Comment:
I don't think you need this variable?

You should be able to access the first/last element of the vector directly:
 - https://en.cppreference.com/w/cpp/container/vector/front
 - https://en.cppreference.com/w/cpp/container/vector/back

Reply:
This is not about accessing the first or last, but the last active window. I don't change the order of the windows inside the vector when a different window becomes active. So I have to keep track of the active window explicitly.

Reply:
Doh! Sorry I missed this.

Agreed, that makes sense.